### DISCUSSION TITLE
Moving from Apache to FrankenPhp : how to manage local virtualhosts and PHP versions ?

### DISCUSSION THREAD

**Discussion by dah-HUB** [php/frankenphp#721] [Q&A] [answered]
Hi,

Actually, I code apps in Laravel,
I wanted to test frankenphp locally; And if its Okay, move the Prod.

How I'm settings up things now : 
* I create virtualhots in /etc/apache2/available-sites/ as this pattern : project name.customer name.local , Ex : crm.lestilleuls.local
* I also add those names to /etc/hosts
* All the the code is located in /var/www/html/ as /var/www/html/PHP version/customer name/project name/

Question 1 : How to manage multiple hosts (apps) in frankenphp ?

I slso use a script to switch PHP versions : 7.4 -> 8.1 -> 8.2 , by Enable/Disable Apache Modules (or services for  FPM) + update-alternatives

Question 2 : How to manage PHP versions in frankenphp ?

Thanks for your Help

**Comment by dunglas** [comment, marked as answer]
FrankenPHP is just a custom build of Caddy, you can manage your virtual hosts as with a standard Caddy installation using site blocks: https://caddyserver.com/docs/caddyfile/concepts#addresses

Supporting multiple PHP installations in the same process is currently not supported. You'll have to start a different instance of FrankenPHP per PHP version.

**Reply by dah-HUB** [reply]
Hi @dunglas 

Thx for your answer,

I'll :
* check the caddy Documentation :+1: 
* Check how to start multiple instances of frankenPhp using different PHP versions :+1: 

have a nice Day

**Reply by SimJoSt** [reply]
Running multiple instances of FrankenPHP creates the need for different ports per instance, correct?
In a production environment on a server with multiple applications using different PHP versions, is not feasible right now. Or at least not without another reverse proxy, like Traefik or Caddy, in front of all the FrankenPHP instances, like described here: https://github.com/php/frankenphp/issues/670#issuecomment-3523842738

Are there future plans to make this possible?

**Reply by henderkes** [reply]
Well, kind of, you can also run multiple instances of php-fpm (like you currently do) and just replace Apache with FrankenPHP. Then use `php_fastcgi` for the applications that require a specific php version and serve all the modern ones with `php_server`.

That's what I do for one very very old forum website that doesn't run on 8.0+.

**Reply by SimJoSt** [reply]
Ok, that makes sense to me. Run all PHP 8.4(+) apps through FrankenPHP directly and then use the "old" method as a fallback for incompatible applications, until they are properly upgraded.
Sounds good for now.

**Reply by SimJoSt** [reply]
In https://github.com/php/frankenphp#standalone-binary it says that standalone binaries contain PHP 8.4 however, there is no mention of which PHP version is contained in the releases in dep packages, etc.
How can I check this reliably?
And what happens if a deb package update in the future moves to a version containing PHP 8.5, but my connected application is not ready for it yet. Will the package upgrade break the setup?

**Reply by henderkes** [reply]
You add the source list of the version you want. 

`curl -fsSL https://files.henderkes.com/install.sh | sudo sh -s 8.5`
`sudo apt install frankenphp`
-> frankenphp with version 8.5

If you want php 8.2, use `sudo sh -s 8.2`

**Reply by henderkes** [reply]
The frankenphp docs will change soon too. Just waiting for someone to review the doc changes with my new repos.

The old debian repository will stay at version 8.4 for a few months, but then be deleted. It already prints a deprecation notice when you install or update.

**Reply by SimJoSt** [reply]
@henderkes where can we follow the status of the docs overhaul? It would be great to see what the future brings.

**Reply by henderkes** [reply]
They already made it live: https://frankenphp.dev/docs/#rpm-packages

**Reply by DavidPetrasek** [reply]
> You add the source list of the version you want.
> 
> `curl -fsSL https://files.henderkes.com/install.sh | sudo sh -s 8.5` `sudo apt install frankenphp` -> frankenphp with version 8.5
> 
> If you want php 8.2, use `sudo sh -s 8.2`

How to achieve that in OpenSUSE? I tried it and it returns: `Error: No supported package manager found (apt, apk, dnf).`

**Reply by henderkes** [reply]
Follow https://rpms.henderkes.com manual approach. Since SUSE uses rpm packages with zypper instead of dnf, the automatic installation won't work.

If you get it working please report back with exact steps, I could add zypper to the list then.

**Reply by DavidPetrasek** [reply]
I don't want to install dnf now... I tried to compile using homebrew and xcaddy/go, but there are some errors

**Reply by henderkes** [reply]
I think you should be able to

```bash
sudo zypper install https://rpm.henderkes.com/static-php-1-1.noarch.rpm
sudo zypper install frankenphp
```

Not sure though because I don't think zypper supports modularity.

**Reply by henderkes** [reply]
And if that doesn't work you could definitely install the packages manually:

```bash
zypper install https://rpm.henderkes.com/x86_64/el9/php-zts-cli-8.5.5-1.el9.x86_64.rpm
zypper install https://rpm.henderkes.com/x86_64/el9/php-zts-embed-8.5.5-1.el9.x86_64.rpm
zypper install https://rpm.henderkes.com/x86_64/el9/frankenphp-1.12.2_85-1.el9.x86_64.rpm
```

But without updates, you'd be better off using docker instead (or installing dnf).

**Reply by DavidPetrasek** [reply]
> I think you should be able to
> 
> ```shell
> sudo zypper install https://rpm.henderkes.com/static-php-1-1.noarch.rpm
> sudo zypper install frankenphp
> ```
> 
> Not sure though because I don't think zypper supports modularity.

This added the `static-php`package but the package `frankenphp` wasn't found.
I wasn't able to make docker work. I'm using OpenSUSE Slowroll

**Reply by DavidPetrasek** [reply]
I was able to compile it with php 8.2: 

- xcaddy and watcher-c were build from source.

- Compilation required to manually copy `/home/linuxbrew/.linuxbrew/Cellar/php@8.2-zts/8.2.30/include/php/Zend` to `/usr/local/include/Zend` and `/watcher/watcher-c/include/wtr/watcher-c.h` to `/usr/local/include/wtr/watcher-c.h`

Compilation required to install these dependencies:
```
libsybdb5
libpq5
libaspell15
libpspell15
```

And I needed to edit the original compile command to:
```
CGO_ENABLED=1 \
XCADDY_GO_BUILD_FLAGS="-ldflags='-w -s' -tags=nobadger,nomysql,nopgx" \
CGO_CFLAGS=$(php-config --includes) \
CGO_LDFLAGS="$(php-config --ldflags) $(php-config --libs) -L$(php-config --prefix)/lib -Wl,-rpath,$(php-config --prefix)/lib -L$(brew --prefix pcre2)/lib" \
/home/david/go/bin/xcaddy build \
    --output frankenphp \
    --with github.com/dunglas/frankenphp/caddy \
    --with github.com/dunglas/mercure/caddy \
    --with github.com/dunglas/vulcain/caddy \
    --with github.com/dunglas/caddy-cbrotli

```

Then I had to: `sudo setcap 'cap_net_bind_service=+ep' /usr/local/bin/frankenphp`

**Reply by DavidPetrasek** [reply]
https://frankenphp.dev/docs/#standalone-binary
Instead of a single standalone binary, what about providing multiple standalone binaries - each with a different PHP version?

**Reply by henderkes** [reply]
No, one with the latest version is enough, it would explore our matrix to far larger than it already is. Using PHP 8.2 is a bad idea anyways because it's quite slow with ZTS.

**Reply by DavidPetrasek** [reply]
Let's say I decide to start working on a new feature "Support multiple PHP installations in the same process" - how many people do you think would be interested in such feature?

**Reply by henderkes** [reply]
That's a very, very, very bad idea. The different versions would 99.99% crash each other with unexpected opcache formats.
